Rank Preference

Utmanarn opened the pull request:
<!-- Guidelines: https://docs.spacestation14.io/en/getting-started/pr-guideline -->

## About the PR
<!-- What did you change? -->
Added a rank preference option to every job to allow players to select their prefered rank level correlating to the ranks you could achieve with playtime.

## Why / Balance
<!-- Discuss how this would affect game balance or explain why it was changed. Link any relevant discussions or issues. -->
It improves the chances for player to RP with different characters, letting some characters stay at certain ranks for RP reasons or to have newly created characters start out as privates.

## Technical details
<!-- Summary of code changes for easier review. -->
Changed the `RankSystem` to take into account the players preference for the ranks while still keeping the requirement for playtime. This works with any number of ranks on the roles.

There is an auto system that allows the player to just ignore this all together and play like nothing has changed! It will always select the max unlocked rank you have.

## Media
<!-- Attach media if the PR makes ingame changes (clothing, items, features, etc). 
Small fixes/refactors are exempt. Media may be used in SS14 progress reports with credit. -->
(This is in the debug mode so all ranks are able to be selected but they are still only valid with playtime)
![Plat_private](https://github.com/user-attachments/assets/955f7a31-94a6-4ee0-92a7-f019c6c6bcd5)
![Auto_Showcase](https://github.com/user-attachments/assets/4491256e-51d5-4b32-baee-a807561d9477)
![Auto_Showcase_Profile](https://github.com/user-attachments/assets/db025335-a105-47b0-bcf6-3da5bb9cd10d)

Ranks get locked if we have not played enough to unlock them.
![Lock_ranks_not_unlocked](https://github.com/user-attachments/assets/81b280aa-844d-48a2-ac86-c7af5bea49a4)

Whops! Our client somehow broke the system and selected a rank that they should not have access to? No worries, the server has the final say in everything and sets the rank to the default base rank if that happens.
![Client_somehow_broke_the_system](https://github.com/user-attachments/assets/4a7fbef8-5d8c-4cef-8fa3-bed2eead3f18)
![Reverts_to_backup_on_client_server_missmatch](https://github.com/user-attachments/assets/12344452-6a90-4e1d-96ae-8ca623fd7bd2)


## Requirements
<!-- Confirm the following by placing an X in the brackets [X]: -->
- [X] I have read and am following the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
- [X] I have added media to this PR or it does not require an ingame showcase.
<!-- You should understand that not following the above may get your PR closed at maintainer’s discretion -->
- [X] By submitting this code and/or assets, I confirm that I either own them or have provided the correct necessary licenses to use and distribute them. I agree to be fully responsible for any legal claims or issues arising from the use of these materials.

**Changelog**
<!-- Add a Changelog entry to make players aware of new features or changes that could affect gameplay.
Make sure to read the guidelines and take this Changelog template out of the comment block in order for it to show up.
Changelog must have a :cl: symbol, so the bot recognizes the changes and adds them to the game's changelog. -->

:cl:
- tweak: Changed the playtime rank system to allow players to select a preference for what type of playtime rank they want to use.

Utmanarn committed:
Work on UI and database

The UI work should hopefully be fine?

The database work is making me cry internally. I have no idea why EF is not working with me.

Utmanarn committed:
Database stuff

Added the database stuff. This stuff was more pain than it was worth.

I can't tell if the modelSnapshot bein changed is good or bad... Oh well... I tried changing my install to the 8.0.0 to match but that did nothing.

Utmanarn committed:
Rank system overhaul

Now the rank system has a preference system! Wooooo! It sucks! But it is my first actual like big thing I do for SS14 so... Yes.

Utmanarn committed:
Localization

Now we have englush!

Dutch-VanDerLinde commented:
Wouldn't it be better to add a rank selection per job

Utmanarn commented:
> Wouldn't it be better to add a rank selection per job

Yeah, I was thinking about that as an option. But I am not sure if loadouts were a good option for that.
I could try to move it over if that is preferred.

Dutch-VanDerLinde commented:
don't need to add it to loadouts. Can have a button next to the job preference which has a list of all the available ranks, buttons are greyed out if the time requirement is wrong

Then at the top of the dropdown there's an auto button (by default) which automatically selects the highest ran

Utmanarn commented:
That is an idea, yes.

I am pretty new to the codebase, so I have yet to figure out most stuff with UI and like most things. So it will probably take me a while to see if I can get it done. But I'll give it my best attempt!

Utmanarn commented:
@Dutch-VanDerLinde
So what do you think of this as a solution?

![bild](https://github.com/user-attachments/assets/526a78a4-f799-44f5-8f0b-0f3e26396e38)

mmails commented:
> @Dutch-VanDerLinde
> 
> So what do you think of this as a solution?
> 
> 
> 
> ![bild](https://github.com/user-attachments/assets/526a78a4-f799-44f5-8f0b-0f3e26396e38)

Make sure this doesn't somehow skip the playtime checks and allow you to use whatever rank

Dutch-VanDerLinde commented:
> @Dutch-VanDerLinde So what do you think of this as a solution?
> 
> ![bild](https://private-user-images.githubusercontent.com/26390071/440020566-526a78a4-f799-44f5-8f0b-0f3e26396e38.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDYyMTM2NTQsIm5iZiI6MTc0NjIxMzM1NCwicGF0aCI6Ii8yNjM5MDA3MS80NDAwMjA1NjYtNTI2YTc4YTQtZjc5OS00NGY1LThmMGItMGYzZTI2Mzk2ZTM4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA1MDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNTAyVDE5MTU1NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTUxNjE0N2Q2M2YzMGYxMmQyZTMzZDU2ZWUxMGE3ZTQ3OTllZWNjNTEwZDU3Yzk5MDNkMTdkMjE1YjljMGYwMTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0._PnwvjFxK3lpvrIPg5vpvrv0qHArKJhRlVCSRuSvo-Y)

Looks great

Dutch-VanDerLinde commented:
Also, for SEA, make sure to make the other ranks selectable too like usual
For greying out the buttons (like a locked job), could it like how loadouts does it with the loadout effects

Utmanarn commented:
I noticed something with option buttons under ScrollContainers look kinda funky. But this is nothing game breaking, nor would I know how to fix it. So unless it is an absolute deal breaker I will leave it for now.

![bild](https://github.com/user-attachments/assets/18b0d9b4-4e3b-4679-9efa-13a7418458f8)

Utmanarn committed:
Reworked rank preferences

Rank preferences has been completely reworked to work with jobs instead of just being generic.

This took way too long and probably is still not done. I have yet to test it because I want to push this before I mess with the stupid database.

Utmanarn committed:
Changed logic and touched database

Changed the logic of the rank system to be more ROBUST. My previous logic was quite frankly flawed and I was working at like 4 AM. BUT NOW! It should be fine. Hopefully. Please.

Also, the database is just evil.

Utmanarn commented:
@Dutch-VanDerLinde Alright! Finally. After much blood, sweat and tears. It is here. The new and improved version that does not require future handholding. (Hopefully)

Still have to do some YAML stuff for validaiton things... But other than that it SHOULD be the final product.

Dutch-VanDerLinde commented:
Good work soldier (I hate database code so much i feel u)

Dutch-VanDerLinde commented on Content.Server/_RMC14/Marines/Roles/Ranks/RankSystem.cs:
```diff
@@ -48,13 +52,42 @@ private void OnPlayerSpawnComplete(PlayerSpawnCompleteEvent ev)
+                // 0 is reserved for auto which skips the whole rank preference system entirely.
+                // (i != jobPrototype.Ranks.Count) makes sure that we at the very least select the default rank
+                // even if the client manages to somehow set their rank preference to something higher than what they have unlocked on the server.
+                if (!skipPreferenceEvaluation && rankPreference != 0 && i != jobPrototype.Ranks.Count)
```
make sure you check o n the server aswell if the rank can be selected

clients could change their rank preference without having the correct time

Dutch-VanDerLinde replied:
nvm, it's good i misread your code

Dutch-VanDerLinde commented on Resources/Locale/en-US/preferences/ui/humanoid-profile-editor.ftl:
```diff
@@ -47,6 +48,11 @@ humanoid-profile-editor-preference-spawn-priority-none = None
 humanoid-profile-editor-preference-spawn-priority-arrivals = Arrivals
 humanoid-profile-editor-preference-spawn-priority-cryosleep = Cryosleep
 
+# Playtime rank preference
+rank-preference-high = High
+rank-preference-medium = Middle
+rank-preference-low = Low
+
```
```suggestion
# Playtime rank preference
rank-preference-high = High
rank-preference-medium = Middle
rank-preference-low = Low

```

Utmanarn replied:
I am working on the YAML stuff rn. Getting that removed and one more which is not needed (further up).

Utmanarn committed:
Updated YAML stuff

Removed unnecessary YAML stuff.

Changed how the CM JobPrototype stuff works. It is no longer default to true for validation. Instead it is true on all jobs that inherit from CM base job and false for all Xeno jobs.

Utmanarn commented:
NOW! It is hopefully the FINAL:tm: version! No more changess unless I messed up that server stuff that @Dutch-VanDerLinde was talking about.

Dutch-VanDerLinde commented on Content.Shared/_RMC14/Marines/Roles/JobPrototype.cs:
```diff
@@ -39,6 +39,12 @@ public sealed partial class JobPrototype : IInheritingPrototype, ICMSpecific
+    /// <summary>
+    /// Should this job have rank preferences appear in the preferences menu?
+    /// </summary>
+    [DataField("setRankPreference")]
```
You don't need datafield names, it'll automatically be assigned

Utmanarn replied:
Alright, I can remove it if requested.

Dutch-VanDerLinde replied:
yes please do its unnessescary

Utmanarn replied:
Done, I also realised another little funny thing while at it.

I had not implemented the functionality of the setting.

Utmanarn committed:
Fixed SetRankPreference

Was going to just remove the string for SetRankPreference but realised I had forgotten to actually implement the functionality past just validation checking.

So it now actually removes the rank preference button from the jobs that are set to false now.

Utmanarn commented:
Please check again to make sure I did not overlook something with the latest changes.

Sometimes even the smallest changes can cause the biggest headaches. But I don't think I missed anything?

If anything, maybe force set the rank to Auto when the preference selection is disabled in case the rank previously had selection enabled? Not sure though if that will ever happen.

Utmanarn committed:
Reworded comment

Made the comment a little more readable and descriptive.

DrSmugleaf reviewed (commented):
Need to think on this, in 13 this only applies to a few roles such as SEA, beyond disabling playtime perks which is supposed to put you at the second lowest rank, but might be fine

Utmanarn committed:
Makes default values not save

We do not want to save default values into the DB. This also fixes a small bug where if you selected auto while already having auto selected the profile system would believe that you changed values when in reality you have not and require you to save your changes.

Dutch-VanDerLinde commented:
merge master to fix tests

Utmanarn committed:
Merge branch 'RMC-14:master' into customise-rank

Dutch-VanDerLinde commented:
Fix da conflicts

Utmanarn committed:
Merge branch 'master' into customise-rank

DrSmugleaf commented on Content.Client/Lobby/UI/HumanoidProfileEditor.xaml.cs:
```diff
@@ -1114,14 +1120,59 @@ public void RefreshJobs()
                         };
                     }
 
+                    var rankOptions = new OptionButton()
```
```suggestion
                    // RMC14
                    var rankOptions = new OptionButton()
```

DrSmugleaf commented on Content.Client/Lobby/UI/HumanoidProfileEditor.xaml.cs:
```diff
@@ -1114,14 +1120,59 @@ public void RefreshJobs()
+                    else
+                    {
+                        rankOptions.Visible = false;
+                    }
```
```suggestion
                    }
                    // RMC14
```

DrSmugleaf committed:
Update Content.Client/Lobby/UI/HumanoidProfileEditor.xaml.cs

DrSmugleaf committed:
Update Content.Client/Lobby/UI/HumanoidProfileEditor.xaml.cs

Aurallianz reviewed (changes requested):
The option for a rank should be greyed out if you do not have that rank.

Try checking if the stroll event is happening, then close the dropdown popup.

Utmanarn commented:
> > > The option for a rank should be greyed out if you do not have that rank.
> > > Try checking if the stroll event is happening, then close the dropdown popup.
> > 
> > 
> > Sorry, I am a bit unsure of what you mean.
> > Ranks you don't have are already greyed out? Unless you meant something else.
> 
> they were not when I was testing, unless im missing something, I should not be able to set my rank to LT Col. if all I have is Major. As for my crappy spelling, if you scroll on the window, close the dropdown.

Do you run it in release mode? The fact that you can select all ranks may be that you run it in debug mode.  Debug mode overrides the time requirements.

As for the scrolling in the dropdown. Does that close the window? It shouldn't do that. It is the base SS14 UI. I will have to try that.

Edit:
Ah, wait. You want it to close the dropdown! Yeah. I guess I can check if I can fix that.

Utmanarn committed:
Merge branch 'master' into customise-rank

Dutch-VanDerLinde commented:
But First Sergeant and Master Sergeant should both be 10 hours, while Master Gunny and Sergeant Major are both unlocked at 70 hours

efzapa commented:
this should probably wait for #8929 to be merged so im not stuck as CMP trying to guess if the Corporal that just woke up is on their first round or secretly ruby playtime

Utmanarn committed:
SEA now has all their ranks

Vixting commented on Content.Server/_RMC14/Marines/Roles/Ranks/RankSystem.cs:
```diff
@@ -48,13 +52,42 @@ private void OnPlayerSpawnComplete(PlayerSpawnCompleteEvent ev)
 
             if (_prototypes.TryIndex<RankPrototype>(rank.Key, out var rankPrototype) && rankPrototype != null)
             {
+                // We can't really havea an enum here to explain the values because there can in theory be infinite values.
```
```suggestion
                // We can't really have a an enum here to explain the values because there can in theory be infinite values.
```

Vixting commented on Content.Client/Lobby/UI/HumanoidProfileEditor.xaml.cs:
```diff
@@ -1114,14 +1120,61 @@ public void RefreshJobs()
+                    // RMC14
+                    var rankOptions = new OptionButton()
+                    {
+                        Name = Loc.GetString("rank-options-button"),
```
this ftl isn't present

Vixting commented on Content.Client/Lobby/UI/HumanoidProfileEditor.xaml.cs:
```diff
@@ -1362,6 +1415,14 @@ private void SetArmorPreference(ArmorPreference newArmorPreference)
 
+        private void SetRankPreference(string jobId, int rankPriority)
+        {
+            var protoJobId = new ProtoId<JobPrototype>(jobId);
```
this isn't used, can be removed.

Vixting commented on Content.Client/Lobby/UI/HumanoidProfileEditor.xaml.cs:
```diff
@@ -38,6 +39,7 @@
 using Robust.Shared.ContentPack;
 using Robust.Shared.Enums;
 using Robust.Shared.Prototypes;
+using Robust.Shared.Toolshed.TypeParsers;
```
not used

Vixting commented on Content.Server/_RMC14/Marines/Roles/Ranks/RankSystem.cs:
```diff
@@ -12,6 +15,7 @@ public sealed class RankSystem : SharedRankSystem
     [Dependency] private readonly PlayTimeTrackingManager _tracking = default!;
     [Dependency] private readonly IPrototypeManager _prototypes = default!;
     [Dependency] private readonly IEntityManager _entityManager = default!;
+    [Dependency] private readonly IServerPreferencesManager _preferences = default!;
```
not used

Vixting reviewed (changes requested):
It would be preferable to store the ranks proto id instead of its positional index.

Instead of this
        private Dictionary<ProtoId<JobPrototype>, int> _rankPreferences = new();

do it like this
        private Dictionary<ProtoId<JobPrototype>, ProtoId<RankPrototype>?> _rankPreferences = new();


storing an index is very brittle instead of just having the rank names

I can help you with making the changes for this if you need it, as there are quite a few things to change.

Utmanarn commented:
> It would be preferable to store the ranks proto id instead of its positional index.
> 
> Instead of this private Dictionary<ProtoId, int> _rankPreferences = new();
> 
> do it like this private Dictionary<ProtoId, ProtoId?> _rankPreferences = new();
> 
> storing an index is very brittle instead of just having the rank names
> 
> I can help you with making the changes for this if you need it, as there are quite a few things to change.

Alright, this was one of my earliest commits and I was quite new to the codebase at the time. Wasn't quite sure exactly how to best implement some things.

If you want to help make the changes you're very welcome to. Otherwise I will see when I get a chance to work on it.

Vixting committed:
removed storing intenger

Vixting commented:
> > It would be preferable to store the ranks proto id instead of its positional index.
> > Instead of this private Dictionary<ProtoId, int> _rankPreferences = new();
> > do it like this private Dictionary<ProtoId, ProtoId?> _rankPreferences = new();
> > storing an index is very brittle instead of just having the rank names
> > I can help you with making the changes for this if you need it, as there are quite a few things to change.
> 
> Alright, this was one of my earliest commits and I was quite new to the codebase at the time. Wasn't quite sure exactly how to best implement some things.
> 
> If you want to help make the changes you're very welcome to. Otherwise I will see when I get a chance to work on it.

I've made the adjustments, if tests pass should be good now.

Vixting reviewed (approved):
LGTM

Aurallianz commented on Content.Client/Lobby/UI/HumanoidProfileEditor.xaml.cs:
```diff
@@ -1114,14 +1119,63 @@ public void RefreshJobs()
                     jobContainer.AddChild(selector);
                     jobContainer.AddChild(loadoutWindowBtn);
+                    jobContainer.AddChild(rankOptions);
                     category.AddChild(jobContainer);
```
```suggestion
                    // RMC14
                    var rankOptions = new OptionButton()
                    {
                        Name = "RankOptionsButton",
                        HorizontalAlignment = HAlignment.Right,
                        VerticalAlignment = VAlignment.Center,
                        Margin = new Thickness(3f, 3f, 0f, 0f)
                    };

                    // index 0 = Auto (null), subsequent entries map 1:1 to job.Ranks in definition order.
                    var rankProtoIds = new List<ProtoId<RankPrototype>?> { null };

                    // if the job does not contain ranks we do not show it (Xenos as an example).
                    if (job.Ranks == null || !job.SetRankPreference)
                    {
                        rankOptions.Visible = false;
                        return;
                    }

                    rankOptions.AddItem("Auto");
                    foreach (var rank in job.Ranks)
                    {
                        if (!_prototypeManager.TryIndex(rank.Key, out RankPrototype? rankPrototype))
                         continue;

                        rankOptions.AddItem(rankPrototype.Name);
                        rankProtoIds.Add(rank.Key);

                        if (rank.Value != null && !_requirements.CheckRoleRequirements(rank.Value, Profile, out _))
                            rankOptions.SetItemDisabled(rankOptions.ItemCount - 1, true);
                    }

                    // If the job only has 1 rank there is nothing to choose.
                    if (rankProtoIds.Count <= 2)
                        rankOptions.Disabled = true;

                    rankOptions.OnItemSelected += args =>
                    {
                        rankOptions.SelectId(args.Id);
                        SetRankPreference(job.ID, rankProtoIds[args.Id]);
                    };

                    _jobPriorities.Add((job.ID, selector));
                    _rankPriorities.Add((job.ID, rankOptions, rankProtoIds));
                    jobContainer.AddChild(selector);
                    jobContainer.AddChild(loadoutWindowBtn);
                    jobContainer.AddChild(rankOptions);
                    category.AddChild(jobContainer);
                    // RMC14
```

Aurallianz commented on Content.Server/_RMC14/Marines/Roles/Ranks/RankSystem.cs:
```diff
@@ -11,7 +13,7 @@ public sealed class RankSystem : SharedRankSystem
 {
     [Dependency] private readonly PlayTimeTrackingManager _tracking = default!;
     [Dependency] private readonly IPrototypeManager _prototypes = default!;
-    [Dependency] private readonly IEntityManager _entityManager = default!;
+    [Dependency] private readonly IServerPreferencesManager _preferences = default!;
```
```suggestion
    [Dependency] private readonly IServerPreferencesManager _preferences = default!;
    [Dependency] private readonly IPrototypeManager _prototypes = default!;
    [Dependency] private readonly PlayTimeTrackingManager _tracking = default!;
```

Aurallianz replied:
alphabetical based on var

Aurallianz commented:
I think github is having a stroke
